### PR TITLE
[RFC] add feasibility checker

### PR DESCRIPTION
**Pull request by joaquimg**
This might end up in a separate package.

It is a possible solution for https://github.com/jump-dev/JuMP.jl/issues/693

### CODE REVIEW

File: test/distances.jl [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+@testset "Set non-default distance" begin
+    for n in 1:3
+        v = rand(n)
+        for s in (MOI.Reals(n), MOI.Zeros(n), MOI.Nonnegatives(n), MOI.Nonpositives(n))
+            @test MOD.distance_to_set(MOD.DefaultDistance(), v, MOI.Reals(n)) ≈ 2 * MOD.distance_to_set(DummyDistance(), v, MOI.Reals(n)) atol=eps(Float64)
+            @test MOD.distance_to_set(MOD.DefaultDistance(), -v, MOI.Reals(n)) ≈ 2 * MOD.distance_to_set(DummyDistance(), -v, MOI.Reals(n)) atol=eps(Float64)
+        end
+    end
+end
```
Comment (matbesancon) [inline test/distances.jl:121]:
```suggestion
end

```

File: test/runtests.jl [hunk 2]
```diff
@@ -4,131 +4,14 @@ using Test
...
-            @test MOD.distance_to_set(MOD.DefaultDistance(), v, MOI.Reals(n)) ≈ 2 * MOD.distance_to_set(DummyDistance(), v, MOI.Reals(n)) atol=eps(Float64)
-            @test MOD.distance_to_set(MOD.DefaultDistance(), -v, MOI.Reals(n)) ≈ 2 * MOD.distance_to_set(DummyDistance(), -v, MOI.Reals(n)) atol=eps(Float64)
-        end
-    end
-end
+include("distances.jl")
 
 include("projections.jl")
+
+include("feasibility_checker.jl")
```
Comment (matbesancon) [inline test/runtests.jl:17]:
```suggestion
include("feasibility_checker.jl")

```

File: src/feasibility_checker.jl [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+end
+
+function constraint_violation_report(model::MOI.ModelLike;
+    varval::Function = variable_primal(model),
+    distance_map::Dict = Dict(),
+    options = FeasibilityCheckerOptions())
+
+    largest, vec = constraint_violation(model, varval = varval, distance_map = distance_map)
+
+    str = " Feasibility report\n\n"
```
Comment (matbesancon) [inline src/feasibility_checker.jl:25]: maybe these using a multiline string:
s = """Feasibility report

...
"""

**Comment by matbesancon** [comment]
Maybe adding docstrings on the functions to see what they are meant to be used for / be doing

File: src/feasibility_checker.jl [hunk 4]
```diff
@@ -0,0 +1,110 @@
+const MOD = MathOptSetDistances
```
Comment (matbesancon) [inline src/feasibility_checker.jl:1]: Is the module name needed here?

File: src/feasibility_checker.jl [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+        return new(
+            names,
+            index,
+        )
+    end
+end
+
+function constraint_violation_report(model::MOI.ModelLike;
+    varval::Function = variable_primal(model),
+    distance_map::Dict = Dict(),
```
Comment (matbesancon) [inline src/feasibility_checker.jl:20]:
```suggestion
    distance_map::AbstractDict = Dict(),
```

File: src/feasibility_checker.jl (hunk 3 above)
Comment (matbesancon) [reply src/feasibility_checker.jl:25]: bump

**Comment by matbesancon** [comment]
re-opening PR to run on actions

File: src/feasibility_checker.jl (hunk 3 above)
Comment (joaquimg) [reply src/feasibility_checker.jl:25]: Agree! I will be back soon to it

**Comment by matbesancon** [comment]
I'm not sure about this conflict resolution, feel free to revert if it caused more damage than good

**Comment by matbesancon** [comment]
ok definitely broke something :grimacing:

**Comment by joaquimg** [comment]
I am writing version 2.0.
Will update soon!

**Comment by joaquimg** [comment]
so...
I moved it here: https://github.com/joaquimg/FeasibilityOptInterface.jl/pull/1
It grew a little too much and went beyond the scope of this package.

**Pull request closed without merging**